### ISSUE TITLE
Reals: Make "Rsqr x" and an alias for "x^2"

### ISSUE THREAD

**Issue by MSoegtropIMC** [coq/coq#10723]
Some developments of the Reals library use a wild mix between `Rsqr x` (aka `x²`) and `x^2`. the field tactic supports (afaikt) only x^2 while most of the lemmas connecting x^2 and sqrt use Rsqr. Trigonometric lemmas use a mix of both. As a result many proofs have to convert between the two forms from time to time. Also proof search is made substantially harder by this.

I would suggest to clean this up by making Rsqr an alias for x^2. The small performance advantage given by Rsqr should not be significant and less important that cleanliness and usability of the library. 

@silene, @thery : can you please comment on this, especially also on the performance implications e.g. in CoqInterval (I always used x^2 there btw.)

**Comment by thery** [comment]
@MSoegtropIMC  @silene making `x ^ 2` convertible to `x * x` would be nice too but I don't know what is the manpower cost of all these refactoring.

**Comment by MSoegtropIMC** [comment]
> making `x ^ 2` convertible to `x * x` would be nice too

Do you mean:
```
Fixpoint pow (r : R) (n : nat) : R :=
  match n with
  | 0%nat => 1
  | 1%nat => r
  | S n' => r * pow r n'
  end.
```

I thought about this - it would be a nice option - not sure how many proofs one would need to change for this.

> I don't know what is the manpower cost of all these refactoring.

I will give it a try - it is also a lot of work to live with this if one works with reals. On the other hand I am not sure what the goals of standard library 2.0 and reals are (@maximedenes ?) . I have once hard that they shall not be part of it - something I would object to, because I think the standard library should cover all high school level math to give a good base for broader use in engineering. Anyway, my guess is that usability improvements for reals in the current standard library is a worthwhile effort.

**Comment by thery** [comment]
>  r * pow r n'

Yes maybe  `pow r n' * r` as * is left associate. With `Arguments pow r n : simpl never.` and a `pow_S` theorem it should be possible to patch all the proofs.

**Comment by silene** [comment]
I have mixed feelings about all that. I agree that it would be better if `x^2` was convertible to `x*x`. But I think the amount of breakage in existing proofs is not worth the effort. (As for making `Rsqr` a notation for `x^2` without changing `x^2` to be `x*x`, I think it is even worse.)

Regarding CoqInterval, reasoning on `x^2` is certainly slower than `x*x`, but it does not matter much in practice, since it is still negligible with respect to other functions.

**Comment by MSoegtropIMC** [comment]
> But I think the amount of breakage in existing proofs is not worth the effort.

I would say this depends on what is planned for standard library 2.0. Leaving it as is is also quite a pain because field doesn't understand `Rsqr` (and I think it shouldn't) but many useful lemmas are using `Rsqr`. Also searching lemmas is really a pain because one has to try `Rsqr` and `_^2`. So I would say it might even take just me less time to change it now than to live with it for the next years.  My guess is that the effort would be about 1...2x the work for the recent change to `field_simplify`, which was quite manageable. Maybe this is optimistic, because Rsqr is much older than `field_simplify` and might be used more widely. On the other hand it might be possible to do the change such that not almost everything becomes incompatible as with `field_simplify`.

What I don't know is if somebody is already working on a replacement for the reals library. I got the impression that the approach is more to evolve the current library to something better, e.g. all the fundamental work by @VincentSe.

**Comment by VincentSe** [comment]
@MSoegtropIMC I don't understand why so many people want to kill the current reals stdlib. It has many theorems implemented already, and its defects can be fixed.

Concerning the estimation of costs of development, it might be cheap advice but my usual opinion is : just do it, you might have good surprises.

**Comment by silene** [comment]
> Leaving it as is is also quite a pain because field doesn't understand `Rsqr` (and I think it shouldn't)

I don't understand why `field` should not know about `Rsqr`. They are both defined in the exact same file, namely `RIneq.v`. It would take only one line to make `field` aware of `Rsqr`.

**Comment by MSoegtropIMC** [comment]
> I don't understand why so many people want to kill the current reals stdlib. It has many theorems implemented already, and its defects can be fixed.

Don't get me wrong, I don't say that the reals lib should be discarded. What I wanted to say is that in case the decision is to keep and further develop the current library, this (the topic of this issue) should be cleaned up. I just don't know what the decision is.

**Comment by MSoegtropIMC** [comment]
> I don't understand why `field` should not know about `Rsqr`.

It is a matter of taste: to me `Rsqr` is simply redundant. But let's distinguish *should* and *could*. In case we decide to keep incompatible `Rsqr` and `pow` functions, it would of cause be better to make field aware of Rsqr. This would solve at least one of the problems.

**Comment by VincentSe** [comment]
You can also do this incrementally :
1) Remove `Rsqr`
2) Compile
3) Fix 10 files
4) Put `Rsqr` back so that the rest compiles
5) Push
6) Go watch a series on Netflix
7) Goto 1

**Comment by thery** [comment]
> I don't understand why `field` should not know about `Rsqr`. They are both defined in the exact same file, namely `RIneq.v`. It would take only one line to make `field` aware of `Rsqr`.

I think the only way to deal with Rsqr is to add it to the preprocessing but then Rsqr will be ring_simplify to x^2 (but could not find where the Add Ring is done to test it).

**Comment by thery** [comment]
I did some experiment with trying to remove `Rsqr` usin @VincentSe' s method but without Step 6.
It is doable https://github.com/thery/coq/commit/f93d11d39a927766523d2532fb86349b6dca36a8
but I am scared it is moving too much the interface to the outside world.
So maybe the best solution is to have ring and field regognized  Rsqr. https://github.com/coq/coq/pull/10734 and  https://github.com/coq/coq/pull/10732 are fixing some issues  to make it possible.

**Comment by VincentSe** [comment]
@thery Nice ! What do you mean by interface and outside world ? You force to use operator `^` instead of function `Rsqr`, that seems simpler and cleaner.

Also, would one of you merge this ?
https://github.com/coq/coq/pull/10713
It's been lonely for 9 days...

**Comment by MSoegtropIMC** [comment]
@thery: thanks for going through it! Did you do a CI run to see the effect on other developments? Maybe you can just do a PR and label it as "needs progress" right away, so that people know it is not ready for merge.

> but I am scared it is moving too much the interface to the outside world.

It is changing it to the better. It is not very intuitive that x^2 and Rsqr x are rather different things. When teaching Coq math to engineers this is the kind of thing people get upset about (like that field_simplify x did result in x/1). People get more annoyed about such little things. If something big is missing, people think, ok this is really something big. But looking at such issues people ask themselves why it has to be like this. Coq improved a lot in this area in the last few years, so let's not stop here.

> So maybe the best solution is to have ring and field regognized Rsqr. #10734 and #10732 are fixing some issues to make it possible.

This is only solving half of the issue. The more time consuming issue is actually search. One (or at least I) can hardly remember which lemmas use x^2 and which use Rsqr, so one has to try both options.

**Comment by thery** [comment]
> What do you mean by interface and outside world

Before `Rsqr x` was untouched by `simpl`, now `x ^ 2` reduces  to `x * (x * 1)`.  This makes proof
scripts really very unstable. 

Some theorem statements have been changed or removed like `Rsqr_pow2`. So the interface
is not the same.

The file `Ranalysis_reg` defines tactics for expressions with pow and Rsqr. I have no idea if 
they behave ok after my changes.

**Comment by MSoegtropIMC** [comment]
> `x ^ 2` reduces to `x * (x * 1)`

Have you experimented with defining pow such that x^2 expands to x*x?

**Comment by maximedenes** [comment]
> Before `Rsqr x` was untouched by `simpl`, now `x ^ 2` reduces to `x * (x * 1)`. This makes proof
> scripts really very unstable.

@thery you should probably make `^` be simplified only when the two arguments are known.
